artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/AddressImpl.java

```diff
@@ -67,17 +69,23 @@ public boolean containsWildCard() {
    }
 
    @Override
-   public List<Address> getLinkedAddresses() {
-      return linkedAddresses;
+   public Collection<Address> getLinkedAddresses() {
+      return linkedAddresses == null ? Collections.emptySet() : linkedAddresses;
```

Comment:
There are a bunch of places where getLinkedAddresses are iterated over, and the collection is not thread safe. We need to double check those are covered by an external sync such that we don't get concurrent mod exceptions.

Reply:
hmm, I think org.apache.activemq.artemis.core.postoffice.PostOffice#updateMessageLoadBalancingTypeForAddress will cause trouble.

Reply:
I see that as well and I'm a bit concern about it too, but I don't think that the initialization will make any difference here...
There is a synchronization problem here. period...
It's weird we've never had issues on this, so I'm going to write a test and check if I can break it. If not, putting a synchronize will just kill performance and the same if we copy the whole linked addresses list (or we use a copy on write data structure too, but on addBinding/removeBinding side)....

Reply:
I added a test for the concurrency issue with updateMessageLoadBalancingTypeForAddress.  The simplest solution is a concurrent set.